core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSession.java

```diff
@@ -0,0 +1,70 @@
+package ai.timefold.solver.core.impl.domain.variable.declarative;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import ai.timefold.solver.core.impl.domain.variable.descriptor.VariableDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.supply.Supply;
+import ai.timefold.solver.core.preview.api.domain.variable.declarative.ShadowVariableSession;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public class DefaultShadowVariableSession<Solution_> implements ShadowVariableSession, Supply {
+    final VariableReferenceGraph<Solution_> graph;
+
+    record EntityVariablePair<Solution_>(VariableDescriptor<Solution_> variableDescriptor, Object entity) {
```

Comment:
Considering what the descriptor is used for, it can easily be replaced by VariableMetaModel.